################################################################################
SHORT DESCRIPTION: 
################################################################################
Changes to support PCI passthrough with the new QEmu without the need for
multiple ioreq servers.

################################################################################
LONG DESCRIPTION: 
################################################################################
Stubdomains need to be given sufficient privilege over the guest which it
provides emulation for in order for PCI passthrough to work correctly.

The capability for multiple device models (the ioreq server patch) was removed
as the current implementation of it is far different from what upstream Xen has
settled on in 4.5.

OXT-118

################################################################################
CHANGELOG 
################################################################################
Authors:
    Signed-off: Eric Chanudet <chanudete@ainfosec.com> (stubdom permissions)
    Signed-off: Rian Quinn <quinnr@ainfosec.com> (ioreq removal)

################################################################################
REMOVAL 
################################################################################

################################################################################
UPSTREAM PLAN
################################################################################

################################################################################
INTERNAL DEPENDENCIES 
################################################################################

################################################################################
PATCHES 
################################################################################
--- a/xen/arch/x86/irq.c
+++ b/xen/arch/x86/irq.c
@@ -2672,8 +2672,36 @@ int allocate_and_map_msi_pirq(struct dom
         irq = index;
         if ( irq == -1 )
         {
-    case MAP_PIRQ_TYPE_MULTI_MSI:
             irq = create_irq(NUMA_NO_NODE);
+            /* Allow stubdomain to deal with this IRQ. */
+            if ( d == current->domain->target )
+            {
+                ret = irq_permit_access(current->domain, irq);
+                if ( ret )
+                    printk(XENLOG_G_ERR "Could not grant stubdom%u access to IRQ%d (error %d)\n",
+                           current->domain->domain_id, irq, ret);
+            }
+	}
+
+        if ( irq < nr_irqs_gsi || irq >= nr_irqs )
+        {
+            dprintk(XENLOG_G_ERR, "dom%d: can't create irq for msi!\n",
+                    d->domain_id);
+            return -EINVAL;
+        }
+
+        msi->irq = irq;
+        break;
+    case MAP_PIRQ_TYPE_MULTI_MSI:
+        irq = create_irq(NUMA_NO_NODE);
+        /* Allow stubdomain to deal with this IRQ. */
+        if ( d == current->domain->target )
+        {
+            ret = irq_permit_access(current->domain, irq);
+            if ( ret )
+                printk(XENLOG_G_ERR "Could not grant stubdom%u access to IRQ%d (error %d)\n",
+                       current->domain->domain_id, irq, ret);
+
         }
 
         if ( irq < nr_irqs_gsi || irq >= nr_irqs )
@@ -2716,8 +2744,20 @@ int allocate_and_map_msi_pirq(struct dom
         {
         case MAP_PIRQ_TYPE_MSI:
             if ( index == -1 )
+            {
+                if ( (d == current->domain->target) &&
+                     irq_deny_access(current->domain, irq) )
+                    printk(XENLOG_G_ERR "dom%d: could not revoke access to IRQ%d.\n",
+                           d->domain_id, irq );
+		destroy_irq(irq);
+            }
+            break;
         case MAP_PIRQ_TYPE_MULTI_MSI:
-                destroy_irq(irq);
+            destroy_irq(irq);
+            if ( (d == current->domain->target) &&
+                  irq_deny_access(current->domain, irq) )
+                printk(XENLOG_G_ERR "dom%d: could not revoke access to IRQ%d.\n",
+                       d->domain_id, irq );
             break;
         }
     }
--- a/xen/arch/x86/physdev.c
+++ b/xen/arch/x86/physdev.c
@@ -164,6 +164,18 @@ int physdev_unmap_pirq(domid_t domid, in
 
     pcidevs_lock();
     spin_lock(&d->event_lock);
+    /* Remove stubdomain's permission on IRQ. */
+    if (d == current->domain->target)
+    {
+        int irq;
+
+        irq = domain_pirq_to_irq(d, pirq);
+        if ( irq <= 0 )
+            printk(XENLOG_G_ERR "dom%d invalid pirq:%d!\n", d->domain_id, pirq);
+        else if ( irq_deny_access(current->domain, irq) )
+            printk(XENLOG_G_ERR "Could not revoke stubdom%u access to IRQ%d.\n",
+                        current->domain->domain_id, irq);
+    }
     ret = unmap_domain_pirq(d, pirq);
     spin_unlock(&d->event_lock);
     pcidevs_unlock();
